API/Services/KoreaderService.cs

```diff
@@ -0,0 +1,52 @@
+using System;
+using System.Threading.Tasks;
+using API.Data;
+using API.DTOs.Koreader;
+using API.Helpers;
+
+namespace API.Services;
+
+#nullable enable
+
+public interface IKoreaderService
+{
+    Task SaveProgress(KoreaderBookDto koreaderBookDto, int userId);
+    Task<KoreaderBookDto> GetProgress(string bookHash, int userId);
+}
+
+public class KoreaderService : IKoreaderService
+{
+    private IReaderService _readerService;
+    private IUnitOfWork _unitOfWork;
+
+    public KoreaderService(IReaderService readerService, IUnitOfWork unitOfWork)
+    {
+        _readerService = readerService;
+        _unitOfWork = unitOfWork;
+    }
+
+    public async Task SaveProgress(KoreaderBookDto koreaderBookDto, int userId)
+    {
+        var file = await _unitOfWork.MangaFileRepository.GetByHash(koreaderBookDto.Document);
+        var userProgressDto = await _unitOfWork.AppUserProgressRepository.GetUserProgressDtoAsync(file.ChapterId, userId);
+        KoreaderHelper.UpdateProgressDto(koreaderBookDto.Progress, userProgressDto);
+        _readerService.SaveReadingProgress(userProgressDto, userId);
+        await _unitOfWork.CommitAsync();
+    }
+
+    public async Task<KoreaderBookDto> GetProgress(string bookHash, int userId)
+    {
+        var file = await _unitOfWork.MangaFileRepository.GetByHash(bookHash);
+        var progressDto = await _unitOfWork.AppUserProgressRepository.GetUserProgressDtoAsync(file.ChapterId, userId);
+        var koreaderProgress = KoreaderHelper.GetKoreaderPosition(progressDto);
+        return new KoreaderBookDto
+        {
+            Document = bookHash,
+            Device_id = "kavita",
+            Device = "kavita",
+            Progress = koreaderProgress,
+            Percentage = 0.5f
```

Comment:
Curious why we are sending a percentage if it's not correct...

Reply:
By what I've seen, Koreader doesn't appear to use this value. I think it could be used by the Koreader progress sync server to display statistics or something. I'll see if I can track it down a little better.

Reply:
So I've been looking at this even more, and [it appears to be only for message boxes.](
https://github.com/koreader/koreader/blob/master/plugins/kosync.koplugin/main.lua#L798)
A [little earlier in the method](https://github.com/koreader/koreader/blob/master/plugins/kosync.koplugin/main.lua#L764) it compares against a previous percentage and a new percentage to see if if is in the same position. If so, it isn't supposed to update the progress in Koreader. However, in testing it doesn't behave that way. Maybe it has to do with direct comparison of a floating point.

Reply:
It looks like it's a user friendly percentage. Can we not use the Progress mapped to a 0-1 float?

Reply:
Someone from Stump did some research on this:
https://github.com/stumpapp/stump/issues/239#issuecomment-2428256328